lib/AST/ASTPrinter.cpp

```diff
@@ -3586,19 +3589,81 @@ class TypePrinter : public TypeVisitor<TypePrinter> {
     return T->hasSimpleTypeRepr();
   }
 
+  /// Add all Clang modules that have been imported into `Mod` to
+  /// `clangModules`.
+  ///
+  /// This also takes into account any modules that are imported transitively
+  /// through public (`@_exported`) imports.
+  /// `clangModules` is a map from each imported Clang module to the
+  /// corresponding Swift module.
+  static void getImportedClangModules(
+      ModuleDecl *Mod, ModuleDecl::ImportFilter filter,
+      llvm::DenseMap<const clang::Module *, ModuleDecl *> &clangModules) {
+    SmallVector<ModuleDecl::ImportedModule, 4> imports;
+    Mod->getImportedModules(imports, filter);
+    for (const auto &import : imports) {
+      if (const clang::Module *clangModule =
+              import.importedModule->findUnderlyingClangModule())
+        clangModules[clangModule] = import.importedModule;
+      getImportedClangModules(import.importedModule,
+                              ModuleDecl::ImportFilterKind::Public,
+                              clangModules);
+    }
+  }
+
+  /// Returns all Clang modules that are visible from `Options.CurrentModule`.
+  /// The returned map associates each visible Clang module with the
+  /// corresponding Swift module.
+  const llvm::DenseMap<const clang::Module *, ModuleDecl *> &
+  getVisibleClangModules() {
+    assert(Options.CurrentModule &&
+           "CurrentModule needs to be set to determine imported Clang modules");
+    if (!VisibleClangModules) {
+      VisibleClangModules.emplace();
+      ModuleDecl::ImportFilter Filter = ModuleDecl::ImportFilterKind::Public;
+      Filter |= ModuleDecl::ImportFilterKind::Private;
```

Comment:
I'm afraid the test case is not testing this subtle difference in filtering for modules imported directly vs. modules a couple hops away.

Reply:
I'm not sure I understand what you mean.

*  `MainModule` imports `HelperModule` privately, and this import should be considered
*  `HelperModule` imports `ForeignA` privately and `ForeignB` publicly, but only the latter should be considered.

So I would think that the test case does cover this. Can you clarify what else you think should be tested?

Reply:
@gribozavr, could you clarify what you meant?

Reply:
@martinboehme is right, the test covers both cases!